### ISSUE TITLE
Custom worker instance groups not referenced in GCE LB ingress

### ISSUE THREAD

**Issue by mbuelte** [poseidon/typhoon#336]
### Environment

* Platform: google-cloud
* OS: container-linux / stable (1855.5.0)
* Typhoon modules: v1.12.2
* Terraform: v0.11.10
+ Terraform provider.ct: v0.3.0
+ Terraform provider.google: v1.19.1
+ Terraform provider.local: v1.1.0
+ Terraform provider.null: v1.0.0
+ Terraform provider.template: v1.0.0
+ Terraform provider.tls: v1.2.0

### Problem

I've setup two worker pools for GCE following the Typhoon documentation and leave out a specific worker configuration in the `variables.tf`, but Teraform is going to create a dedicated worker instance named "$CLUSTER_PREFIX-worker-$ID". I did not noticed this behavior prior v1.12.2. If I delete this node afterwards it is recreated some time later by "../InstanceGroupManagers/xxx-worker-group". And now the real problem; This single instance is in its own "Instance group" named " $CLUSTER_PREFIX-worker-group" where both ingresses (http/https) are pointing to. Also the health check of the load balancer has a problem with this instance it shows a unhealthy state that leads to the problem that the public nginx-ingress isn't reachable.	

***terraform plan**
```shell
...
google_compute_instance_template.worker: Refreshing state... (ID: xxx-small-static-worker-20181103093325568000000003)
google_compute_instance_template.worker: Refreshing state... (ID: xxx-small-preemptible-worker-20181103093325414000000002)
google_compute_region_instance_group_manager.workers: Refreshing state... (ID: xxx-small-static-worker-group)
google_compute_region_instance_group_manager.workers: Refreshing state... (ID: xxx-worker-group)
...
```

**variables.tf**
```hcl
module "google-cloud" {
  source = "git::https://github.com/poseidon/typhoon//google-cloud/container-linux/kubernetes?ref=v1.12.2"

  providers = {
    google   = "google.default"
    local    = "local.default"
    null     = "null.default"
    template = "template.default"
    tls      = "tls.default"
  }

  # Google Cloud
  cluster_name  = "${var.gcloud_cluster_name}"
  region        = "${var.gcloud_region}"
  dns_zone      = "${var.gcloud_dns_zone}"
  dns_zone_name = "${var.gcloud_dns_zone_name}"

  # General configuration
  cluster_domain_suffix = "xxx.local"
  ssh_authorized_key = "${var.gcloud_ssh_authorized_key}"
  asset_dir          = xxx/.secrets/clusters/xxx"

  # Nodes
  os_image           = "coreos-stable"
  controller_count   = 2
  controller_type    = "n1-standard-1"
}
```

**worker-pool.tf**
```hcl
module "worker-pool-small-static" {
  source = "git::https://github.com/poseidon/typhoon//google-cloud/container-linux/kubernetes/workers?ref=v1.12.2"

  providers = {
    google = "google.default"
  }

  # Google Cloud
  region       = "${var.gcloud_region}"
  network      = "${module.google-cloud.network_name}"
  cluster_name = "${var.gcloud_cluster_name}"

  # Configuration
  name               = "${var.gcloud_cluster_name}-small-static"
  kubeconfig         = "${module.google-cloud.kubeconfig}"
  ssh_authorized_key = "${var.gcloud_ssh_authorized_key}"

  # Optional
  count        = 1
  machine_type = "n1-standard-1"
  os_image     = "coreos-stable"
  preemptible  = false
}

module "worker-pool-small-preemptible" {
  source = "git::https://github.com/poseidon/typhoon//google-cloud/container-linux/kubernetes/workers?ref=v1.12.2"

  providers = {
    google = "google.default"
  }

  # Google Cloud
  region       = "${var.gcloud_region}"
  network      = "${module.google-cloud.network_name}"
  cluster_name = "${var.gcloud_cluster_name}"

  # Configuration
  name               = "${var.gcloud_cluster_name}-small-preemptible"
  kubeconfig         = "${module.google-cloud.kubeconfig}"
  ssh_authorized_key = "${var.gcloud_ssh_authorized_key}"

  # Optional
  count        = 2
  machine_type = "n1-standard-1"
  os_image     = "coreos-stable"
  preemptible  = true
}
```

### Desired Behavior

I would expect that all "custom" workers from the pools are in the generic instance group "$CLUSTER_PREFIX-worker-group" or at least referenced in the load balancer backend configuration. It was completely fine prior v1.12.2 that the additional worker wasn't created. Maybe I'm not using the worker pool correctly or the documentation is outdated. I also tried to define "worker_count = 0" in `variables.tf` but this didn't change anything.
Let me know if more informations are required.

**Comment by dghubble** [comment]
Typhoon GCP clusters consist of a desired number of controllers and a managed instance group of workers (default `worker_count` is 1) that are homogeneous (e.g. workers have the same type, preemptibility, etc.). For convenience, each cluster comes with load balancer forwarding rules forwarding to workers running a healthy Ingress Controller. Apply an Ingress controller (see [nginx-ingress](https://typhoon.psdn.io/addons/ingress/#google-cloud) addon) to use the feature, otherwise its perfectly fine and expected there won't be ingress backends. Nothing about this design has changed in recent releases and Typhoon clusters on AWS and Azure provide the same model.

[Workers pools](https://typhoon.psdn.io/advanced/worker-pools/) (advanced) are for more skilled users who wish to attach additional (different!) managed instance groups (or ASGs on AWS or SS's on Azure) of workers to a cluster for certain workloads (e.g. a mix of `non-preemptible` and `preemptible`, different machine types, GPUs, etc.). Attached worker pools aren't part of the included ingress backend service (its not generally plausible to do so, especially consistently on different clouds). Depending on the complexity of attached worker pools, if you want to use the included ingress load balancer, you'll have to think about where you want ingress controllers to be run.

Looking at your usage, first, there's no reason for that `static-small` worker pool. Its no different from the workers in the cluster's default managed instance group of workers (same machine_type, region, os_image, etc.) Delete it and instead just set `worker_count` to desired. So then you'll have a cluster with a mix of non-premeptible and prepemtible (via the worker pool) workers.

Next, you'd typically just need to ensure some ingress controllers run on the non-preemptible workers (either run enough you're sure there will be some on non-premeptible workers, use a daemonset, or select with node selectors or taints). If you decide you also want to serve ingress traffic from worker pools (e.g. your premptible workers), you should create your own load balancer instead of using the convenient included one. Each GCP worker pool has an output [variable](https://github.com/poseidon/typhoon/blob/master/google-cloud/container-linux/kubernetes/workers/outputs.tf#L1) `instance_group` intended to allow creating load balancers with any combination of worker pools.

**Comment by mbuelte** [comment]
Hey, thank you for your detailed answer. Both worker pools were just tests. I know that the simple one makes no sense in this configuration. Some time ago I evaluated Typhoon version 1.12.1 with exact the same configurations, but probably some other dependencies versions of Terraform plugins and the single worker wasn‘t created at all and the pool worker were in the loadbalancer instance group after the installation of the nginx-ingress addon. So I’m just wondering what happens. Isn‘t it possible after all to stop Typhoon to create worker?

In conclusion a easy solution for my problem is just to add every custom worker instance group to the included load balancer setup and live with the single one that creates Typhoon for me?

**Comment by mbuelte** [comment]
Alright thanks again. Probably someone else here deleted the managed default worker instance group and added the custom ones to the lb backend. And for whatever reasons Terraform didn’t recognize a change. Guess this issue can be marked as solved.
Just a short additional question: Is there any disadvantage to let every default worker be preemtible? I can‘t find one, especially if the cluster configuration will drain this node before it’s termination. Will Typhoon take care that there is at least one worker node running or start a new one shortly after the last worker is gone?

**Comment by dghubble** [comment]
> Is there any disadvantage to let every default worker be preemtible?

Well yes. Every worker may be preempted at the same time and not replaced for an arbitrary amount of time (though very unlikely). Typhoon is not involved here. Managed instance groups are a Google Cloud concept. They aim to maintain a desired number of nodes. With `premeptible=true`, you're agreeing to pay a lower price, but Google doesn't guarantee they will have capacity for you (e.g. if they're overloaded). Be sure you're cool with the risk. https://cloud.google.com/compute/docs/instances/preemptible
